packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,239 @@
+
+        for (const vault of vaults) {
+            for (const item of vault.items) {
+                const passwordFields = item.fields.filter((field) => field.type === FieldType.Password);
+
+                if (passwordFields.length === 0) {
+                    continue;
+                }
+
+                let isReused = false;
+                let isWeak = false;
+                // let isCompromised = false;
+
+                for (const passwordField of passwordFields) {
+                    const passwordHash = await sha1(passwordField.value);
+
+                    // Perform reused audit
+                    if (Object.keys(firstMatchPerPasswordHash).includes(passwordHash)) {
+                        isReused = true;
+
+                        // Don't add the same item twice to the list, if there are more than one reused password fields in it
+                        if (!reusedPasswordItemIds.has(item.id)) {
+                            reusedPasswords.push({ item, vault });
+                            reusedPasswordItemIds.add(item.id);
+                        }
+
+                        // TODO: Save audit match boolean in field
+                        // passwordField.auditResult = { ...(passwordField.auditResult || {}), isReused };
+
+                        // Also tag the first matching item as reused, once
+                        const firstMatch = firstMatchPerPasswordHash[passwordHash];
+                        if (!reusedPasswordItemIds.has(firstMatch.item.id)) {
+                            reusedPasswords.push({ item: firstMatch.item, vault: firstMatch.vault });
+                            reusedPasswordItemIds.add(firstMatch.item.id);
+
+                            // TODO: Save audit match boolean in field
+                            // firstMatch.passwordField.auditResult = { ...(firstMatch.passwordField.auditResult || {}), isReused };
+                        }
+                    }
+
+                    firstMatchPerPasswordHash[passwordHash] = { item, vault, passwordField };
+
+                    // Perform weak audit
+                    if (isPasswordWeak(passwordField.value)) {
+                        isWeak = true;
+
+                        // Don't add the same item twice to the list, if there are more than one reused password fields in it
+                        if (!weakPasswordItemIds.has(item.id)) {
+                            weakPasswords.push({ item, vault });
+                            weakPasswordItemIds.add(item.id);
+                        }
+
+                        // TODO: Save audit match boolean in field
+                        // passwordField.auditResult = { ...(passwordField.auditResult || {}), isWeak };
+                    }
+
+                    // TODO: Perform compromised audit
+                }
+
+                // TODO: Save audit match booleans in item
+                // item.auditResult = { lastAudited: new Date(), isReused, isWeak, isCompromised };
```

Comment:
@MaKleSoft and this one for the item itself (so we can filter by).

Reply:
FYI: You'll also need to call `Vault::updateItem`:

```ts
item.auditResults = ...
item.lastAudited = new Date();
vault.updateItem(item);
```

For updating invidual items, there is also the convenience method `app.updateItem`, but that will also trigger `app.save()` and `app.synchronize()` on every call, so it shouldn't be used for bulk updates:

```ts
app.updateItem(item, { auditResults: ..., lastAudited: new Date() })
```

Reply:
I don't see a `vault.updateItem()` nor anything similar on `vault` or `item` , only `app.save()`. Is that not enough, or should I use something else?